lib/propolis/src/vmm/mem.rs

```diff
@@ -1063,6 +1065,61 @@ impl<T: Copy + FromBytes> Iterator for GuestData<MemMany<'_, T>> {
     }
 }
 
+/// A 52-bit physical page number, i.e., the ordinal index of a 4 KiB page of
+/// guest memory.
+#[derive(Clone, Copy, Ord, PartialOrd, Eq, PartialEq)]
+pub(crate) struct Pfn(u64);
+
+impl std::fmt::Debug for Pfn {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "Pfn({:x})", self.0)
+    }
+}
+
+impl std::fmt::Display for Pfn {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        core::fmt::LowerHex::fmt(&self.0, f)
+    }
+}
+
+impl From<Pfn> for u64 {
+    fn from(value: Pfn) -> Self {
+        value.0
+    }
+}
+
+impl Pfn {
+    /// Creates a new PFN wrapper for the supplied physical page number. Returns
+    /// `None` if the page number cannot correctly be shifted to form the
+    /// corresponding 64-bit address.
+    pub(crate) fn new(pfn: u64) -> Option<Self> {
+        if pfn > (PAGE_MASK as u64 >> PAGE_SHIFT) {
```

Comment:
nitpicky, feel free to ignore me: personally I like to define max values for things like this as a public associated constant in case it needs to be referenced in other code --- though this is more important in a public library API, and since this is crate private, we could always just go and add a constant for `PAGE_MASK as u64 >> PAGE_SHIFT` later if we need it. so...maybe that doesn't matter.

Reply:
I've left this as-is. As you mention, this is internal-only, and I suspect this is a case where callers aren't especially likely to care about the maximum possible 4K PFN (they're more likely to care about, say, whether a PFN is within the bounds of the guest's physical memory, but that depends on how much guest memory there is).